File: packages/core/src/pdf/pageDecor.ts
import { intersectsBox, smoothCurves, type Pt } from './geometry.js'
import type { PdfPage } from './pdfPage.js'
import type { PagePoint } from '../layout/planPreview.js'
import {
  mapHeightMm,
  mapWidthMm,
  type MapPage,
  type PageFrame,
  type PaperSpec,
} from '../layout/pageLayout.js'
/** Millimetres to PostScript points. */
export function mm(value: number): number {
  return (value * 72.0) / 25.4
}

export interface PageDecorOptions {
  readonly frame: PageFrame
  readonly paper: PaperSpec
  readonly attribution: string
  readonly title: string | null
  /**
   * The walk, one polyline per leg, in page-frame metres. Null or empty draws nothing,
   * which is the default and what every page looked like before this was an option.
   */
  readonly track?: readonly (readonly PagePoint[])[] | null
}

/**
 * Violet.
 *
 * SCAN25 spends red and orange on roads, blue on water, green on woodland, brown on relief —
 * and **magenta on GR waymarking**, which is exactly the kind of path a walker's GPX follows.
 * Violet is the nearest hue that is legible on all of those and confusable with none, and the
 * white halo under it settles the rest.
 */
export const TRACK_RGB: readonly [number, number, number] = [0.35, 0.0, 0.75]

const TRACK_WIDTH_PT = 1.4
const TRACK_HALO_PT = 3.2

/**
 * Both passes are translucent, so the SCAN25 path under the trace stays readable.
 *
 * The halo is the weaker of the two on purpose: opaque white it erased the very footpath
 * the walker is following, which is the one thing the trace is meant to point at. At these
 * values the violet still separates from woodland and hillshade, and the dashes of a
 * footpath count through it.
 */
export const TRACK_ALPHA = 0.5
export const TRACK_HALO_ALPHA = 0.25

/**
 * Everything printed on top of the map: a north arrow that accounts for the page rotation,
 * the scale bar, the page number and the neighbour hints.
 *
 * The Lambert-93 kilometre grid used to be drawn here too. It was removed on request: it
 * put a blue lattice over every square centimetre of an already dense map. The scale bar is
 * still an exact kilometre, so it remains the ruler test's measuring device.
 *
 * The GPX trace is drawn only when a `track` is passed. It defaults to off: the trace's job
 * is to decide where the pages go, and a printed line over 1:25000 detail hides as much as
 * it explains. Asked for on 2026-09-17.
 */
export class PageDecor {
  private readonly frame: PageFrame
  private readonly paper: PaperSpec
  private readonly attribution: string
  private readonly title: string | null
  private readonly track: readonly (readonly PagePoint[])[]

  private readonly mapX: number
  private readonly mapY: number
  private readonly mapW: number
  private readonly mapH: number
  /** Ground metres per PostScript point at the printing scale. */
  private readonly metresPerPt: number

  constructor(options: PageDecorOptions) {
    this.frame = options.frame
    this.paper = options.paper
    this.attribution = options.attribution
    this.title = options.title
    this.track = options.track ?? []

    this.mapX = mm(this.paper.safeMarginMm)
    this.mapY = mm(this.paper.safeMarginMm + this.paper.footerMm)
    this.mapW = mm(mapWidthMm(this.paper))
    this.mapH = mm(mapHeightMm(this.paper))
    this.metresPerPt = (this.paper.scaleDenominator * 25.4) / 72.0 / 1000.0
  }

  draw(canvas: PdfPage, page: MapPage, total: number): void {
    this.drawTrack(canvas, page)
    canvas.setStroke(0.0, 0.0, 0.0)
    canvas.setLineWidth(0.6)
    canvas.strokeRect(this.mapX, this.mapY, this.mapW, this.mapH)

    this.drawFooter(canvas, page, total)
  }

  // --- the walk ----------------------------------------------------------------------

  /**
   * The trace: one smooth translucent polyline per leg, clipped to the map area.
   *
   * Three things it is not. It is not a chain of straight segments — a centripetal
   * Catmull-Rom spline runs through the GPX points so the printed line has no kink at each
   * fix. It is not opaque — both passes are translucent so the path on the map underneath
   * still reads. And it is not clipped by hand any more — the clip is a PDF path, which is
   * what lets a curve cross the page edge without being cut into segments first.
   *
   * Still drawn twice: a white halo, then the line. One pass over dark forest or a
   * hillshaded slope disappears into it.
   */
  private drawTrack(canvas: PdfPage, page: MapPage): void {
    if (this.track.length === 0) return
    const rect = page.rect
    const toPt = (p: PagePoint): Pt => [
      this.mapX + (p.u - rect.uMin) / this.metresPerPt,
      this.mapY + (p.v - rect.vMin) / this.metresPerPt,
    ]

    const visible = this.track
      .map((leg) => leg.map(toPt))
      .filter(
        (leg) =>
          leg.length >= 2 &&
          // The clip would hide it anyway, but a leg on another page must leave the content
          // stream byte for byte as it was, or "no track" and "a track off this page" would
          // produce different files.
          intersectsBox(leg, this.mapX, this.mapY, this.mapX + this.mapW, this.mapY + this.mapH),
      )
    if (visible.length === 0) return

    canvas.save()
    canvas.clipRect(this.mapX, this.mapY, this.mapW, this.mapH)
    canvas.setRoundJoins()
    for (const [colour, width, alpha] of [
      [[1.0, 1.0, 1.0] as const, TRACK_HALO_PT, TRACK_HALO_ALPHA],
      [TRACK_RGB, TRACK_WIDTH_PT, TRACK_ALPHA],
    ] as const) {
      canvas.setStroke(colour[0], colour[1], colour[2])
      canvas.setLineWidth(width)
      canvas.setAlpha(alpha)
      for (const leg of visible) {
        canvas.moveTo(leg[0]![0], leg[0]![1])
        for (const [c1x, c1y, c2x, c2y, x, y] of smoothCurves(leg)) {
          canvas.curveTo(c1x, c1y, c2x, c2y, x, y)
        }
        canvas.strokePath()
      }
    }
    canvas.restore()
  }

  // --- overlays ----------------------------------------------------------------------

  /**
   * North, and the turn the reader has to undo, at the right-hand end of the scale bar's row.
   *
   * It used to be a 13 x 17 mm white box in the top-right corner of the map — 220 mm² of
   * SCAN25 painted over on every page, and on a rotated page that corner is as likely to hold
   * the walk as any other. It followed the neighbour hints into the footer on 2026-09-18,
   * where it also balances the scale bar at the other end of the same row.
   *
   * The needle still has to point anywhere on the compass, so this is the one piece of footer
   * furniture that needs height rather than a baseline.
   */
  private drawNorthArrow(canvas: PdfPage, cy: number): void {
    const [nu, nv] = this.frame.northOnPage()
    const rotation = `${Math.round((360.0 - this.frame.angleDeg) % 360.0)}°`
    const needle = mm(2.2)
    const gap = mm(1.6)
    const width =
      needle * 2 + gap + canvas.textWidth('N', 7.0, true) + gap + canvas.textWidth(rotation, 6.0)

    const cx = this.mapX + this.mapW - width + needle
    const tipX = cx + nu * needle
    const tipY = cy + nv * needle
    // Perpendicular, for the arrow head.
    const px = -nv * mm(0.9)
    const py = nu * mm(0.9)

    canvas.setStroke(0.0, 0.0, 0.0)
    canvas.setLineWidth(0.8)
    canvas.line(cx - nu * needle, cy - nv * needle, tipX, tipY)
    canvas.setFill(0.0, 0.0, 0.0)
    canvas.fillPolygon([
      [tipX, tipY],
      [tipX - nu * mm(1.8) + px, tipY - nv * mm(1.8) + py],
      [tipX - nu * mm(1.8) - px, tipY - nv * mm(1.8) - py],
    ])
    // Cap height is about 0.7 em, so half of it is what puts a label on the needle's axis.
    canvas.text(cx + needle + gap, cy - mm((7.0 * 0.7 * 25.4) / 72 / 2), 7.0, 'N', true)
    canvas.textRight(this.mapX + this.mapW, cy - mm((6.0 * 0.7 * 25.4) / 72 / 2), 6.0, rotation)
  }

  private drawFooter(canvas: PdfPage, page: MapPage, total: number): void {
    const baseline = mm(this.paper.safeMarginMm + 1.5)
    const line = baseline + mm(5.0)
    const number = `Page ${page.number} / ${total}`
    canvas.setFill(0.0, 0.0, 0.0)
    canvas.text(this.mapX, line, 9.0, number, true)

    const attribution = `1:${this.paper.scaleDenominator} · ${this.attribution}`
    canvas.textRight(this.mapX + this.mapW, line, 7.0, attribution)

    let cursor = this.mapX + canvas.textWidth(number, 9.0, true) + mm(4.0)
    cursor += this.drawNeighbourHints(canvas, page, cursor, line)
    if (this.title !== null) {
      const room = this.mapX + this.mapW - canvas.textWidth(attribution, 7.0) - mm(4.0) - cursor
      const title = elide(canvas, this.title, 8.0, room)
      if (title !== null) canvas.text(cursor, line, 8.0, title)
    }
    this.drawScaleBar(canvas, this.mapX, baseline)
    // High enough that the needle clears the safe margin below and the footer's text row
    // above: it is the one piece of furniture down here that is taller than a line of type.
    this.drawNorthArrow(canvas, baseline + mm(1.1))
  }

  /**
   * The pages the map carries on onto, in the footer rather than on the map.
   *
   * They used to be white tabs pinned to the four map edges — which is exactly where the
   * trace leaves the page, so each one covered the detail a walker needs most. The footer
   * is the only space on the sheet that is off the map *and* inside the safe margin, so
   * that is where they went (2026-09-18). Direction survives as a filled triangle, because
   * WinAnsi encoding has no arrow characters.
   *
   * @returns the width used, so the title can start after it.
   */
  private drawNeighbourHints(canvas: PdfPage, page: MapPage, x: number, y: number): number {
    const arrow = mm(2.2)
    const gap = mm(1.4)
    let at = x
    const hint = (numbers: readonly number[], dirX: number, dirY: number): void => {
      if (numbers.length === 0) return
      const cy = y + mm(0.9)
      const half = arrow / 2
      canvas.setFill(0.0, 0.0, 0.0)
      canvas.fillPolygon([
        [at + half + dirX * half, cy + dirY * half],
        [at + half - dirX * half - dirY * half, cy - dirY * half + dirX * half],
        [at + half - dirX * half + dirY * half, cy - dirY * half - dirX * half],
      ])
      at += arrow + mm(0.8)
      const label = `p. ${numbers.join(', ')}`
      canvas.text(at, y, 7.0, label, true)
      at += canvas.textWidth(label, 7.0, true) + gap * 2
    }
    // Reading order round the page: back, up, on, down.
    hint(page.neighbours.left, -1.0, 0.0)
    hint(page.neighbours.up, 0.0, 1.0)
    hint(page.neighbours.right, 1.0, 0.0)
    hint(page.neighbours.down, 0.0, -1.0)
    return at === x ? 0 : at - x
  }

  private drawScaleBar(canvas: PdfPage, x: number, y: number): void {
    const totalM = 1000.0
    const segments = 5
    const barW = totalM / this.metresPerPt
    const segW = barW / segments
    const h = mm(1.4)
    for (let i = 0; i < segments; i++) {
      if (i % 2 === 0) canvas.setFill(0.0, 0.0, 0.0)
      else canvas.setFill(1.0, 1.0, 1.0)
      canvas.fillRect(x + i * segW, y, segW, h)
    }
    canvas.setStroke(0.0, 0.0, 0.0)
    canvas.setLineWidth(0.4)
    canvas.strokeRect(x, y, barW, h)
    canvas.setFill(0.0, 0.0, 0.0)
    canvas.text(x, y - mm(2.6), 6.0, '0')
    canvas.textRight(x + barW, y - mm(2.6), 6.0, '1 km')
  }
}

/**
 * `value` cut to fit `room`, with an ellipsis, or null when even the ellipsis will not fit.
 *
 * The footer is one row and the neighbour hints took some of it, so a long trace name has
 * to give way rather than run into the attribution on the right.
 */
function elide(canvas: PdfPage, value: string, size: number, room: number): string | null {
  if (room <= 0) return null
  if (canvas.textWidth(value, size) <= room) return value
  for (let length = value.length - 1; length > 0; length--) {
    const cut = `${value.slice(0, length).trimEnd()}…`
    if (canvas.textWidth(cut, size) <= room) return cut
  }
  return null
}

File: core/src/main/kotlin/io/gpxtoign/core/pdf/PageDecor.kt
package io.gpxtoign.core.pdf

import io.gpxtoign.core.layout.Layout
import io.gpxtoign.core.layout.MapPage
import io.gpxtoign.core.layout.PaperSpec
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.hypot
import kotlin.math.roundToInt

/** Millimetres to PostScript points. */
internal fun mm(value: Double) = value * 72.0 / 25.4

/**
 * Everything printed on top of the map: the Lambert-93 kilometre grid, a north arrow that
 * accounts for the page rotation, the scale bar, the page number and the neighbour hints.
 *
 * The GPX trace itself is deliberately never drawn — it only decided where the pages go.
 */
class PageDecor(
    private val layout: Layout,
    private val paper: PaperSpec,
    private val attribution: String,
    private val title: String?,
) {
    private val mapX = mm(paper.safeMarginMm)
    private val mapY = mm(paper.safeMarginMm + paper.footerMm)
    private val mapW = mm(paper.mapWidthMm)
    private val mapH = mm(paper.mapHeightMm)

    /** Ground metres per PostScript point at the printing scale. */
    private val metresPerPt = paper.scaleDenominator * 25.4 / 72.0 / 1000.0

    fun draw(canvas: PdfPage, page: MapPage, total: Int) {
        val rect = page.rect
        fun toPt(u: Double, v: Double) = Pair(
            mapX + (u - rect.uMin) / metresPerPt,
            mapY + (v - rect.vMin) / metresPerPt,
        )

        drawKilometreGrid(canvas, page, ::toPt)
        canvas.setStroke(0.0, 0.0, 0.0)
        canvas.setLineWidth(0.6)
        canvas.strokeRect(mapX, mapY, mapW, mapH)

        drawNorthArrow(canvas)
        drawNeighbourTabs(canvas, page)
        drawFooter(canvas, page, total)
    }

    // --- kilometre grid ----------------------------------------------------------------

    private fun drawKilometreGrid(
        canvas: PdfPage, page: MapPage, toPt: (Double, Double) -> Pair<Double, Double>,
    ) {
        val rect = page.rect
        val corners = listOf(
            layout.toL93(rect.uMin, rect.vMin), layout.toL93(rect.uMax, rect.vMin),
            layout.toL93(rect.uMin, rect.vMax), layout.toL93(rect.uMax, rect.vMax),
        )
        val xMin = corners.minOf { it.x }
        val xMax = corners.maxOf { it.x }
        val yMin = corners.minOf { it.y }
        val yMax = corners.maxOf { it.y }
        val reach = hypot(rect.width, rect.height)

        canvas.setStroke(0.15, 0.35, 0.75)
        canvas.setLineWidth(0.3)
        canvas.setFill(0.15, 0.35, 0.75)

        for (km in ceil(xMin / GRID_M).toInt()..floor(xMax / GRID_M).toInt()) {
            val x = km * GRID_M
            val mid = (yMin + yMax) / 2.0
            drawGridLine(canvas, toPt, x, mid - reach, x, mid + reach, "${km}")
        }
        for (km in ceil(yMin / GRID_M).toInt()..floor(yMax / GRID_M).toInt()) {
            val y = km * GRID_M
            val mid = (xMin + xMax) / 2.0
            drawGridLine(canvas, toPt, mid - reach, y, mid + reach, y, "${km}")
        }
    }

    private fun drawGridLine(
        canvas: PdfPage, toPt: (Double, Double) -> Pair<Double, Double>,
        x1: Double, y1: Double, x2: Double, y2: Double, label: String,
    ) {
        val (u1, v1) = layout.toPage(io.gpxtoign.core.geo.L93(x1, y1))
        val (u2, v2) = layout.toPage(io.gpxtoign.core.geo.L93(x2, y2))
        val (px1, py1) = toPt(u1, v1)
        val (px2, py2) = toPt(u2, v2)
        val clipped = Segments.clip(px1, py1, px2, py2, mapX, mapY, mapX + mapW, mapY + mapH)
            ?: return
        canvas.line(clipped[0], clipped[1], clipped[2], clipped[3])
        // Label at whichever end sits lower on the page, nudged inside the frame.
        val (lx, ly) = if (clipped[1] <= clipped[3]) {
            clipped[0] to clipped[1]
        } else {
            clipped[2] to clipped[3]
        }
        val tx = lx.coerceIn(mapX + 1.0, mapX + mapW - mm(7.0))
        val ty = ly.coerceIn(mapY + 1.0, mapY + mapH - mm(4.0))
        val width = canvas.textWidth(label, 5.0) + 2.0
        canvas.setFill(1.0, 1.0, 1.0)
        canvas.fillRect(tx, ty, width, 6.5)
        canvas.setFill(0.15, 0.35, 0.75)
        canvas.text(tx + 1.0, ty + 1.5, 5.0, label)
    }

    // --- overlays ----------------------------------------------------------------------

    private fun drawNorthArrow(canvas: PdfPage) {
        val boxW = mm(13.0)
        val boxH = mm(17.0)
        val x = mapX + mapW - boxW - mm(2.0)
        val y = mapY + mapH - boxH - mm(2.0)

        canvas.setFill(1.0, 1.0, 1.0)
        canvas.fillRect(x, y, boxW, boxH)
        canvas.setStroke(0.0, 0.0, 0.0)
        canvas.setLineWidth(0.5)
        canvas.strokeRect(x, y, boxW, boxH)

        val cx = x + boxW / 2.0
        val cy = y + mm(7.5)
        val (nu, nv) = layout.northOnPage()
        val len = mm(5.5)
        val tipX = cx + nu * len
        val tipY = cy + nv * len
        val tailX = cx - nu * len
        val tailY = cy - nv * len
        // Perpendicular, for the arrow head.
        val px = -nv * mm(1.8)
        val py = nu * mm(1.8)

        canvas.setLineWidth(0.8)
        canvas.line(tailX, tailY, tipX, tipY)
        canvas.setFill(0.0, 0.0, 0.0)
        canvas.fillPolygon(
            listOf(
                tipX to tipY,
                tipX - nu * mm(3.5) + px to tipY - nv * mm(3.5) + py,
                tipX - nu * mm(3.5) - px to tipY - nv * mm(3.5) - py,
            ),
        )
        canvas.textCentered(cx, y + mm(1.5), 7.0, "N", bold = true)
        val rotation = ((360.0 - layout.angleDeg) % 360.0).roundToInt()
        canvas.textCentered(cx, y + boxH - mm(3.0), 5.0, "$rotation°")
    }

    /**
     * White tabs on each edge naming the page that carries the map on. The arrow is a
     * triangle rather than a glyph, because WinAnsi encoding has no arrow characters.
     */
    private fun drawNeighbourTabs(canvas: PdfPage, page: MapPage) {
        fun tab(numbers: List<Int>, x: Double, y: Double, dirX: Double, dirY: Double) {
            if (numbers.isEmpty()) return
            val label = "p. " + numbers.joinToString(", ")
            val arrow = mm(3.0)
            val w = canvas.textWidth(label, 7.0, bold = true) + arrow + mm(3.0)
            val h = mm(4.5)
            canvas.setFill(1.0, 1.0, 1.0)
            canvas.fillRect(x - w / 2, y - h / 2, w, h)
            canvas.setStroke(0.0, 0.0, 0.0)
            canvas.setLineWidth(0.4)
            canvas.strokeRect(x - w / 2, y - h / 2, w, h)

            val ax = x - w / 2 + mm(1.0) + arrow / 2
            val half = arrow / 2
            canvas.setFill(0.0, 0.0, 0.0)
            canvas.fillPolygon(
                listOf(
                    ax + dirX * half to y + dirY * half,
                    ax - dirX * half - dirY * half to y - dirY * half + dirX * half,
                    ax - dirX * half + dirY * half to y - dirY * half - dirX * half,
                ),
            )
            canvas.text(x - w / 2 + mm(1.0) + arrow + mm(1.0), y - mm(1.1), 7.0, label, bold = true)
        }
        val cx = mapX + mapW / 2
        val cy = mapY + mapH / 2
        tab(page.neighbours.up, cx, mapY + mapH - mm(3.0), 0.0, 1.0)
        tab(page.neighbours.down, cx, mapY + mm(3.0), 0.0, -1.0)
        tab(page.neighbours.left, mapX + mm(16.0), cy, -1.0, 0.0)
        tab(page.neighbours.right, mapX + mapW - mm(16.0), cy, 1.0, 0.0)
    }

    private fun drawFooter(canvas: PdfPage, page: MapPage, total: Int) {
        val baseline = mm(paper.safeMarginMm + 1.5)
        canvas.setFill(0.0, 0.0, 0.0)
        canvas.text(mapX, baseline + mm(5.0), 9.0, "Page ${page.number} / $total", bold = true)
        title?.let {
            canvas.text(
                mapX + canvas.textWidth("Page ${page.number} / $total  ", 9.0, bold = true),
                baseline + mm(5.0), 8.0, it,
            )
        }
        canvas.textRight(
            mapX + mapW, baseline + mm(5.0), 7.0,
            "1:${paper.scaleDenominator} · $attribution · grille Lambert-93 (1 km)",
        )
        drawScaleBar(canvas, mapX, baseline)
    }

    private fun drawScaleBar(canvas: PdfPage, x: Double, y: Double) {
        val totalM = 1000.0
        val segments = 5
        val barW = totalM / metresPerPt
        val segW = barW / segments
        val h = mm(1.4)
        for (i in 0 until segments) {
            if (i % 2 == 0) canvas.setFill(0.0, 0.0, 0.0) else canvas.setFill(1.0, 1.0, 1.0)
            canvas.fillRect(x + i * segW, y, segW, h)
        }
        canvas.setStroke(0.0, 0.0, 0.0)
        canvas.setLineWidth(0.4)
        canvas.strokeRect(x, y, barW, h)
        canvas.setFill(0.0, 0.0, 0.0)
        canvas.text(x, y - mm(2.6), 6.0, "0")
        canvas.textRight(x + barW, y - mm(2.6), 6.0, "1 km")
    }

    private companion object {
        const val GRID_M = 1000.0
    }
}
